autogen/middleware/base.py

```diff
@@ -0,0 +1,226 @@
+
+    @property
+    def _is_async(self) -> bool:
+        ...  # pragma: no cover
+
+    @property
+    def _is_method(self) -> bool:
+        ...  # pragma: no cover
+
+    @property
+    def _middlewares(self) -> List[Middleware]:
+        ...  # pragma: no cover
+
+    def __call__(self, *args: Any, **kwargs: Any) -> Any:
+        ...  # pragma: no cover
+
+
+def hookable(f: F) -> Hookable:
+    """Decorator to make a function hookable.
+
+    You can attach middlewares to a hookable function using the `add_middleware` or `set_middlewares` methods.
+
+    Args:
+        f: The function to make hookable.
+
+    Returns:
+        A hookable version of the function.
+    """
+
+    @wraps(f)
+    def h_sync(*args: Any, **kwargs: Any) -> Any:
+        return h_sync._chained_call(*args, **kwargs)  # type: ignore[attr-defined]
+
+    @wraps(f)
+    async def h_async(*args: Any, **kwargs: Any) -> Any:
+        return await h_async._chained_call(*args, **kwargs)  # type: ignore[attr-defined]
+
+    # this is to reduce number of mypy errors below, although we don't care about errors here
+    h: Hookable = h_async if inspect.iscoroutinefunction(f) else h_sync  # type: ignore[assignment]
+
+    h._origin = f  # type: ignore[misc]
+    h._chained_call = f  # type: ignore[misc]
+
+    # if the name of the first argument is `self` or `cls`, then we assume
+    # this is a method, otherwise it is a function
+    sig = inspect.signature(f)
+    first_param_name = list(sig.parameters.keys())[0]
+    h._is_method = first_param_name in ("self", "cls")  # type: ignore[misc]
+
+    # if async, then we need to await the result
+    h._is_async = inspect.iscoroutinefunction(f)  # type: ignore[misc]
+
+    h._middlewares = []  # type: ignore[misc]
+
+    return h
+
+
+def match_caller_type(*, callee: Callable[..., Any], caller: Callable[..., Any]) -> Callable[..., Any]:
+    """Match the caller type of two functions.
+
+    If both `callee` and `caller` are both async or both sync functions, then `callee` is returned.
+    If `callee` is sync and `caller` is async, then a wrapper is returned that calls `callee` and awaits the result.
+    If `callee` is async and `caller` is sync, a wrapper is returned that waits for `callee` and returns the result.
+
+    Args:
+        callee: The source function.
+        caller: The destination function.
+
+    Returns:
+        A function that matches the caller type of `caller`.
+
+    Raises:
+        TypeError: If `callee` is async and `caller` is sync.
+    """
+    # return the callee if already the same type
+    if inspect.iscoroutinefunction(caller) == inspect.iscoroutinefunction(callee):
+        return callee
+
+    loop = asyncio.get_running_loop()
+
+    if inspect.iscoroutinefunction(callee):
+
+        @wraps(callee)
+        def _sync_callee(*args: Any, **kwargs: Any) -> Any:
+            coro = callee(*args, **kwargs)
+            future = asyncio.run_coroutine_threadsafe(coro, loop)
```

Comment:
Today I learned that this asyncio.run_coroutine_threadsafe allows wrapping of async with sync. Thanks!